_posts/2018-10-25-Rust-1.30.md

```diff
@@ -0,0 +1,355 @@
+with `serde_json`, you could
+
+```rust
+#[derive(Serialize, Deserialize, Debug)]
+struct Pet {
+    name: String,
+}
+```
+
+And convert a `Pet` to and from JSON.
+
+Rust expands on this by adding the ability to define two other kinds of
+advanced macros, "attribute-like procedrual macros" and "function-like
+procedural macros."
+
+Attribute-like macros are similar to custom derive macros, but instead of
+generating code for `#[derive]`, they allow you to create new, custom
+attributes of your own. They're also more flexible; derive only works for
+structs and enums; attributes can go on other places as well, like functions.
+As an example of using an attribute-like macro, you might have something like
+this when using a web application framework:
+
+```
+#[route(GET, "/")]
+fn index() {
+```
+
+This `#[route]` attribute would be defined by the framework itself, as a
+procedural macro. Its signature would look like this:
+
+```
+#[proc_macro_attribute]
+pub fn route(attr: TokenStream, item: TokenStream) -> TokenStream {
+```
+
+Here, we have two input `TokenStreams`; the first is for the contents of the
+attribute itself, that is, the `GET, "/"` stuff. The second is the body of the
+thing the attribute is attached to, in this case, `fn index() {}` and the rest
+of the function's body.
+
+Function-like macros define macros that look like function calls. For
+example, an `sql!` macro:
+
+```rust
+let sql = sql!(SELECT * FROM posts WHERE id=1);
+```
+
+This macro would parse the SQL statement inside of it and check that it's
+syntactically correct. This macro would be defined like this:
+
+```
+#[proc_macro]
+pub fn sql(input: TokenStream) -> TokenStream {
+```
+
+This is similar to the derive macro's signature: we get in the tokens that
+are inside of the parentheses, and return the code we wanted to generate.
+
+### `use` and macros
+
+You can now [use `use` to bring macros into scope][externmacro]. For example, 
+to use `serde-json`'s `json` macro, you used to write:
+
+```rust
+#[macro_use]
+extern crate serde_json;
+
+let john = json!({
+    "name": "John Doe",
+    "age": 43,
+    "phones": [
+    "+44 1234567",
+    "+44 2345678"
+    ]
+});
+```
+
+But now, you'd write
+
+```rust
+extern crate serde_json;
+
+use serde_json::json;
+
+let john = json!({
+    "name": "John Doe",
+    "age": 43,
+    "phones": [
+    "+44 1234567",
+    "+44 2345678"
+    ]
+});
+```
+
+This brings macros more in line with other items, and removes the need for
+`macro_use` annotations.
+
+[externmacro]: https://github.com/rust-lang/rust/pull/50911/
+[Rust 1.15]: https://blog.rust-lang.org/2017/02/02/Rust-1.15.html
+
+## Module system improvements
+
+The module system has long been a pain point of new Rustaceans; several of
+its rules felt awkward in practice. These changes are the first steps we're
+taking to make the module system feel more straightforward.
+
+[`mod.rs` files are now optional][optionalmod]. Imagine we had a `foo`
```

Comment:
This one is un-stabilized again in https://github.com/rust-lang/rust/pull/55315 due to [issues](https://github.com/rust-lang/rust/issues/55094).

Reply:
I think you can mention tool attributes instead (like `#[rustfmt::skip]`), they are well hidden in https://github.com/rust-lang/rust/blob/master/RELEASES.md#misc in the release notes.

Reply:
FYI, I've just opened up #55331.  The `tool_lints` feature doesn't actually land until 1.31, even though the compiler warns and recommends the new syntax.

Reply:
@jtgeibel 
The release notes talk about tool attributes (`#[rustfmt::skip]`) rather than tool lints (`#[allow(clippy::something)]`).
Tool attributes were indeed stabilized on 1.30.